week5/Calc.hs

```diff
@@ -0,0 +1,95 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module Calc where
+
+import ExprT
+import Parser
+import StackVM
+import Data.Maybe
+import qualified Data.Map as M
+
+-- Exercise 1
+eval :: ExprT -> Integer
+eval (ExprT.Lit x) = x
+eval (ExprT.Add x y) = eval x + eval y
+eval (ExprT.Mul x y) = eval x * eval y
+
+-- Exercise 2
+evalStr :: String -> Maybe Integer
+evalStr x = case parseExp ExprT.Lit ExprT.Add ExprT.Mul x of
+            Just y -> Just (eval y)
+            _ -> Nothing
+
+-- Exercise 3
+class Expr a where
+  lit :: Integer -> a
+  mul :: a -> a -> a
+  add :: a -> a -> a
+
+instance Expr ExprT where
+  lit = ExprT.Lit
+  mul = ExprT.Mul
+  add = ExprT.Add
+
+-- Exercise 4
+instance Expr Integer where
+  lit = id
+  mul = (*)
+  add = (+)
+
+instance Expr Bool where
+  lit = (>0)
+  mul = (&&)
+  add = (||)
+
+newtype MinMax = MinMax Integer deriving (Eq, Show)
+newtype Mod7   = Mod7 Integer deriving (Eq, Show)
+
+instance Expr MinMax where
+  lit = MinMax
+  mul (MinMax x) (MinMax y) = MinMax (if x < y then x else y)
+  add (MinMax x) (MinMax y) = MinMax (if x > y then x else y)
+
+instance Expr Mod7 where
+  lit = Mod7
+  mul (Mod7 x) (Mod7 y) = Mod7(mod (x*y) 7)
+  add (Mod7 x) (Mod7 y) = Mod7(mod (x+y) 7)
+
+-- Exercise 5
+instance Expr StackVM.Program where
+  lit x = [StackVM.PushI x]
+  add x y = x ++ y ++ [StackVM.Add]
+  mul x y = x ++ y ++ [StackVM.Mul]
+
+compile:: String -> Maybe Program
+compile = parseExp lit add mul
+
+--  Exercise 6
+type MapExpr = M.Map String Integer -> Maybe Integer
+
+class HasVars a where
+  var :: String -> a
+
+instance HasVars MapExpr where
+    var = M.lookup
+
+data VarExprT = Lit Integer
+                | Var String
+                | Add VarExprT VarExprT
+                | Mul VarExprT VarExprT
+
+instance Expr MapExpr where
+  lit x _ = Just x
+  add f g = \x -> case (isNothing (f x) || isNothing (g x)) of
+                  True  -> Nothing
+                  _     -> Just (fromJust (f x) + fromJust (g x))
```

Comment:
It's actually an anti-pattern in Haskell to use `isNothing/isJust + fromJust` function because it combines two potentially problematic approaches:

1. Bool blindness
2. Partial functions

Such code is error-prone and hard to reason about. Instead it's better to do pattern-matching:

```haskell
case f x of
    Nothing -> ...
    Just a -> case g x of
        Nothing -> ...
        Just b -> ...
```

This is a very common way to write programs. But this manual pattern-matching looks awkward and inconvenient. That's why there are tools in Haskell to write such code nice (it can be rewritten much simple using Monads and standard functions) but that's later topic 🙂

Reply:
Understood... revised with current suggestion for now.